### PR TITLE
AML-37 Restrict optuna for diminishing results

### PR DESCRIPTION
**Pull request by tstreule**
This PR includes:
- Restrict optuna for diminishing results
    - Prune when cv score is bad to previous trials
    - Stop trial when...
        a. `N` trials have been pruned consecutively, and
        b. `M > N` trials have been pruned in total
- Prettify a few docstrings
- Adjust import order according to coding guidelines
- Errata in grid search testing (remove preliminary `return`)
- Adjust `merge_logs` function in utils

### CODE REVIEW

**Review by nielsuit227 — CHANGES_REQUESTED** [review]

File: Amplo/GridSearch/OptunaGridSearch.py [hunk 1]
```diff
@@ -157,16 +191,87 @@ def objective(self, trial):
...
+
+    # Decide whether trial is good enough
+    threshold = 0.1
+    if best_is_0:
+        return scoring_value > 0 - threshold if greater_is_better else scoring_value < 0 + threshold
+    elif best_is_1:
+        assert greater_is_better
+        return scoring_value > 1 - threshold
+    else:
+        raise NotImplementedError(f'Cannot decide whether scoring is good or bad: {scoring}')
```
[the remark this answers is not in the document]
Comment (tstreule) [reply Amplo/GridSearch/OptunaGridSearch.py:277]: With `neg_log_loss` I agree.

What do you mean with "we miss the goal here"? Assuming you refer to the threshold of 0.1, I think you're missing the whole picture. While `trial_is_good_enough` is one way of pruning, we also prune when a CV score has no improvements to previous trials (see variable `cv_interrupted`).

[the remark this answers is not in the document]
Comment (nielsuit227) [reply Amplo/GridSearch/OptunaGridSearch.py:277]: `trial_is_good_enough` looks whether the score is reasonable to rely on, which is problem dependent, and trial independent. Therefore, a study without any improvements for half an hour while we still have 100s of candidates or hours left may or may not be pruned based on the difficulty of the problem. 
Instead, this should reflect on the last `x` trials, see how long there hasn't been an improvement (or even how marginal the improvements are) and decide on that whether it's useful to continue.

File: Amplo/Utils/io.py [hunk 2]
```diff
@@ -106,48 +100,92 @@ def merge_logs(path_to_folder: Union[str, Path], target: str = 'labels') -> Tupl
...
+
+    Returns
+    -------
+    metadata : dict
+        Dictionary whose keys depict the file id (integer) and each value contains a dictionary with
+        ``folder``, ``file``, ``full_path`` and ``last_modified`` key.
+    """
+    # Checks
+    assert Path(path_to_folder).is_dir(), 'Invalid path to directory'
+    assert Path(path_to_folder).exists(), 'Directory does not exist'
```
[the remark this answers is not in the document]
Comment (tstreule) [reply Amplo/Utils/io.py]: Nope. See errata in commit 415085ddcca229d863ab8a8392b4b5fc8c60d683

File: Amplo/GridSearch/OptunaGridSearch.py [hunk 3]
```diff
@@ -120,9 +136,14 @@ def fit(self, x, y):
         # Store
         self.x, self.y = x, y
 
+        # Early stopping callback
+        study_stop_cb = StopWhenTrialKeepBeingPrunedCallback(15)
```
[the remark this answers is not in the document]
Comment (tstreule) [reply Amplo/GridSearch/OptunaGridSearch.py]: Good point! `IsThisABetterNameForACallback`?
see commit 6fb2b1df834b3443aecd4fbdaa447d464048139a

File: Amplo/GridSearch/OptunaGridSearch.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tstreule) [reply Amplo/GridSearch/OptunaGridSearch.py:299]: Do I understand you correctly that you want to compare the scores even across multiple models?

For example:
- Top 3 XGBoost models had scores: 0.8, 0.75, 0.74
- Top 3 RF models had scores: 0.79, 0.76, 0.73
Therefore, when the next model, let's say CatBoost, with the given parameter set has a score of `0.79`, then we are happy and say that the trial is good enough.

Like that?

[the remark this answers is not in the document]
Comment (nielsuit227) [reply Amplo/GridSearch/OptunaGridSearch.py:299]: This whole ticket had one goal: Avoid running endless trials without finding improvements. 

Often found 500 trials where the optimum was trial 37. Pruning can mitigate that problem significantly. Both inside the CV but also the whole study when too many trials are crap. 

I didn't even think of using previous trials of different models on the same dataset, but that absolutely makes sense!

[the remark this answers is not in the document]
Comment (tstreule) [reply Amplo/GridSearch/OptunaGridSearch.py]: Please have a look at commit a2739c0d404a3ff62bbf035fb38333bbcbfb48c6

File: Amplo/GridSearch/OptunaGridSearch.py [hunk 4]
```diff
@@ -157,16 +178,158 @@ def objective(self, trial):
...
+
+        # Pruning
+        curr_trial_mean = np.mean(list(trial.intermediate_values.values()))
+        if study.direction == optuna.study.StudyDirection.MAXIMIZE:
+            return curr_trial_mean < threshold_max
+        elif study.direction == optuna.study.StudyDirection.MINIMIZE:
+            return curr_trial_mean > threshold_min
+        else:
+            raise RuntimeError(
+                'Pruning with a threshold is arbitrary when the study direction is undefined.')
```
Comment (nielsuit227) [inline Amplo/GridSearch/OptunaGridSearch.py:258]: We still use scorers, this will always be maximize. It's not even an option in this class....

**Review by nielsuit227 — CHANGES_REQUESTED** [review]

File: Amplo/GridSearch/OptunaGridSearch.py (hunk 4 above)
Comment (tstreule) [reply Amplo/GridSearch/OptunaGridSearch.py:255]: Let's get rid of it then

Comment (tstreule) [reply Amplo/GridSearch/OptunaGridSearch.py:255]: Let's get rid of it then

**Pull request merged**